AS::TimeZone.all/zones_map provide access only to MAPPING

brainopia opened the pull request:
there is a different result depending on sequence of commands
1.

``` ruby
chicago = ActiveSupport::TimeZone['America/Chicago']
ActiveSupport::TimeZone.all.include?(chicago) # => true
```

2.

``` ruby
ActiveSupport::TimeZone.all
chicago = ActiveSupport::TimeZone['America/Chicago']
ActiveSupport::TimeZone.all.include?(chicago) # => false
```

I propose to fix an edge-case when `#[]` can accidentally populate `#all` timezones, therefore in both cases the result would become `false`.

Commit:
Simplify AS::TimeZone

rafaelfranca commented on activesupport/lib/active_support/values/time_zone.rb:
```diff
@@ -223,8 +223,10 @@ def all
       def zones_map
         @zones_map ||= begin
-          MAPPING.each_key {|place| self[place]} # load all the zones
-          @lazy_zones_map
```
Can we get ride of `@lazy_zones_map`?

brainopia commented:
@rafaelfranca indeed, `@lazy_zones_map` can be dropped. There is no noticeable difference with or without it (the main hit could be `AS::TimeZone.all` but it's already cached). I was planning this change in a separate commit, but after your comment decided to bundle a couple more things.

I've dropped `#zones_map`:
- it was undocumented
- had no use in rails and I've found none outside
- all the information can be easily accessed from `#all`, the only prettier use-case is access by name, but `#[]` is better suited for this role.

And I've simplified `#utc_offset` since there is no use of `@current_period` and tzinfo, current_period always present (otherwise there would be an exception not nil from these methods).

rafaelfranca commented:
It was undocumented but part of public API, we can't remove it without deprecation unfortunately.

brainopia commented:
@rafaelfranca I've returned `zones_map` back but with deprecation, would it be acceptable?

dijonkitchen commented:
Is this still relevant or should it be closed for being stale?

rafaelfranca commented:
This is going to be a breaking change so it is better if we document this in the changelog. Do you want to rebase this branch and add a CHANGELOG entry?
